Доработал автотесты для конвертера + мердж новой ветки Никиты

kosedo03 opened the pull request.

nmkazantsev committed:
first commit

nmkazantsev committed:
converter works

nmkazantsev committed:
converter test passed

nmkazantsev committed:
converter test pased

nmkazantsev committed:
add request to printer

nmkazantsev committed:
add static to docker

nmkazantsev committed:
remove base.py

nmkazantsev committed:
fix base.py

nmkazantsev committed:
change file naming

nmkazantsev committed:
befire check out to another branch

nmkazantsev committed:
before checkout to another branch

nmkazantsev committed:
some fixes

nmkazantsev committed:
push

nmkazantsev committed:
fix converter
add runnbale from dict selction

nmkazantsev committed:
add runnnable from dict select

nmkazantsev committed:
delete CONTENT_TYPES add EXTENTIONS

nmkazantsev committed:
блекнул

nmkazantsev committed:
change converter logic (add abstract class)

nmkazantsev committed:
add middleware-check of file size

nmkazantsev committed:
change error code

nmkazantsev committed:
add aio http

nmkazantsev committed:
some fixes
add java to Docker

nmkazantsev committed:
delete versions form requirements.txt

nmkazantsev committed:
clean dependencies

nmkazantsev committed:
fix everything but not handelrs

nmkazantsev committed:
fix handlers

nmkazantsev committed:
break post router

nmkazantsev committed:
everythin works except unit tests

nmkazantsev committed:
test passed

kosedo03 committed:
Тесты на винде не запускались, через Никиту попросил это, чтобы понять, что связано с ОС. Исправил авт. тесты, добавив форму в запрос, все теперь работает. Подчистил код и в README написал инструкцию по тестам и libre.

kosedo03 committed:
Тесты на винде не запускались, через Никиту попросил это, чтобы понять, что связано с ОС. Исправил авт. тесты, добавив форму в запрос, все теперь работает. Подчистил код и в README написал инструкцию по тестам и libre.

kosedo03 committed:
Допилил тесты. Подчистил код и дописал инструкцию.

Добавил форму в запрос.

grigoriev-semyon commented on README.md:
```diff
@@ -20,6 +20,23 @@ foo@bar:~$ pip install -r requirements.txt
+## Тестирование и LibreOffice
+
+1) Установка LibreOffice должна производиться в Docker'e, иначе при выполнении 
+строки через shell возникнут проблемы с доступом к диреткории /static
```
typo

grigoriev-semyon commented on README.md:
```diff
@@ -20,6 +20,23 @@ foo@bar:~$ pip install -r requirements.txt
+1) Установка LibreOffice должна производиться в Docker'e, иначе при выполнении 
+строки через shell возникнут проблемы с доступом к диреткории /static
+
+2) При запуске автоматических тестирований на Windows необходимо в utils/convertable.py
+заменить во всех обращениях к shell через run команды 
+на соответсвующие ОС, указать полную директорию, 
+которую можно удобно узнать в пайчарме, а также полный путь на вашем компьютере до soffice.exe, который указывается в двойных кавычках 
+**(!!Но затем вернуть всё к исходному!!)** :
```
без изменений кода никак?

mrerro replied:
добавте код на распознавание OC, и также поиск soffice.exe во всех стандартных местах где он может быть.

kosedo03 replied:
Добавил поиск по Program Files для винды, правда, время тестов увеличилось

grigoriev-semyon commented on file_converter/routes/base.py:
```diff
@@ -0,0 +1,44 @@
+@app.exception_handler(aiohttp.client_exceptions.ClientConnectorError)
+async def not_found_error(request: Request, exc: aiohttp.client_exceptions.ClientConnectorError):
+    raise HTTPException(
+            500,
```
не найдено - 404 а не 500

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,43 @@
+            pass
+
+
+class Input(BaseModel):
+    to_ext: str
+
```
используется?

kosedo03 replied:
Нет

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,43 @@
+router = APIRouter()
+
+
+async def main(url: str, data: dict):
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url, data=data) as response:
+            pass
```
main непонятно

kosedo03 replied:
Убрал, вместе с printer_URL

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,43 @@
+    if file.filename.split(".")[-1] not in settings.EXTENTIONS:
+        raise HTTPException(
+            415,
+            f'Only {", ".join(settings.EXTENTIONS)} files allowed, but {file.content_type} received',
```
content-type и extension это разные вещи
мы используем только расширение

grigoriev-semyon commented on file_converter/settings.py:
```diff
@@ -0,0 +1,30 @@
+    CONVERT_TYPES: List[str] = ['pdf']
+    MAX_SIZE: int = 5000000  # Максимальный размер файла в байтах
+    STATIC_FOLDER: str = DirectoryPath | None
+    PRINTER_URL: str = "https://yandex.ru"
```
?

kosedo03 replied:
Переписал, теперь просто JSON возвращаем

grigoriev-semyon commented on file_converter/utils/convertable.py:
```diff
@@ -0,0 +1,83 @@
+        await run(f"cd static; libreoffice --headless --convert-to pdf {file_name}")
+
+
+def random_str(n):
+    alph = string.ascii_letters
+    s = ""
+    for i in range(n):
+        s = f"{s}{alph[random.randint(0, len(alph) - 1)]}"
+    return s
```
можно написать в одну строку

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,43 @@
+
+    result = await convert(file, to_ext, settings.STATIC_FOLDER)
+    if not await check_pdf_ok(result):
+        await run(f"rm {result}")
```
кажется неправильная команда

grigoriev-semyon commented on file_converter/utils/convertable.py:
раздели файл на несколько 
большой уже

grigoriev-semyon commented on tests/conftest.py:
```diff
@@ -0,0 +1,9 @@
+from file_converter.routes.base import app
+
+
+@pytest.fixture(scope='session')
```
скоуп не нужен

kosedo03 committed:
pull_r fixes

Исправил + добавил response Schema

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,37 @@
+    if not await check_pdf_ok(result):
+        raise (HTTPException(413, "file corrupted"))
+
+    return {"status": "okay", "file_dir": f"{cd}{result}"}
```
отдавай сразу полную ссылку абсолютную

kosedo03 replied:
Добавил в Settings ROOT_PATH, который из енв берет ссыль

grigoriev-semyon commented on file_converter/utils/check_pdf.py:
```diff
@@ -0,0 +1,13 @@
+    try:
+        PdfFileReader(io.BytesIO(f))
+        return True
+    except Exception as e:
```
здесь точно except Exception? иили можно сузить класс ошибок? может SyntaxError?

mrerro replied:
Чем тут будет полезней сужение типов ошибок?

grigoriev-semyon replied:
> Чем тут будет полезней сужение типов ошибок?

Ну сейчас оно на все говорит 415 или 413 не помню
Не важно что кинуло ошибку - io или PdfFileReader, какую ошибку, может там не broken pdf?
Пусть лучше будет кидать 500, чем я буду смотреть логи и думать почему не грузит норм файл

mrerro replied:
Вообще зачем эта функция тут? у нас идет проверка на pdf в print-api это функция скопирована оттуда, зачем она конвертеру? проверять что выдал либреофис?
Второе предложи тогда как гарантированно узнать какие ошибки можно получить за ранее? Когда фиксили проверку pdf в print-api то можно словить PyPdfError если это не pdf или ValueError если файл бит. Но выборка битости мала и есть шанс словить что то еще и как гарантировать что ты не пропустишь какую-нибудь дич?

grigoriev-semyon replied:
Вообще действительно функцию можно отсюда убрать так как она уже есть в print-api

насчет второго хз, скорее всего только путем исследований так как насколько я помню дока там очень плохая(не написано что какие ошибки кидает)

kosedo03 replied:
Соглашусь, раз в принтере проверка на пдф уже есть, то убрано

grigoriev-semyon commented on file_converter/utils/check_pdf.py:
```diff
@@ -0,0 +1,13 @@
+async def check_pdf_ok(full_file: str):
+    if not exists(full_file):
+        return False
+    f = open(full_file, "rb").read()
```
почему не в менеджере контекста?

kosedo03 replied:
Убрано вовсе

grigoriev-semyon commented on get_dir.py:
```diff
@@ -0,0 +1,20 @@
+import os
+import platform
+
+if platform.system() == "Windows":
+
+    slash = '\\'
+    for root, dirs, files in os.walk('C:\Program Files'):
+        if 'soffice.exe' in files:
+            libre_path = os.path.join(root, 'soffice.exe')
+    command = f' && "{libre_path}" --headless --convert-to pdf '
+
+
+else:
+    slash = '/'
+    command = 'cd static; libreoffice --headless --convert-to pdf '
+
+ext_d = os.path.abspath(" ")
+cd = ext_d[:-2]+slash+'static'+slash
+
+"""Gets a cd for soffice.exe on Windows and /static directory independently from the OS"""
```
а почему это не иинкапсулировано в класс Convertable?

grigoriev-semyon replied:
это значит что этот код можно выполнить один раз при инициализации класса Convertable.
Можно попробовать возвращать не строки
Можно в класс записать Callable 
типа инициализируешь класс выполняя вот эту функцию да и записываешь в классовую переменную Callable.
там будет уже прописанная команда, но не будет файла
и чет типа command(filename) будет

kosedo03 replied:
Переделал в класс и функцию в функции

grigoriev-semyon commented on get_dir.py:
```diff
@@ -0,0 +1,20 @@
+
+if platform.system() == "Windows":
+
+    slash = '\\'
+    for root, dirs, files in os.walk('C:\Program Files'):
+        if 'soffice.exe' in files:
```
ты пишешь про нестандартные пути 
а почему ты ищешь не по всем файлам?
зачем потом менять код?

mrerro replied:
тут смотря на сколько надо углубиться, но минимум надо проверить ProgramFiles, ProgramFiles(x86), %localappdata%/Programs, текущую директорию - вдруг portable подложили. И использовать переменные среды windows. Диск `C` не обязательно системный.

grigoriev-semyon commented on README.md:
```diff
@@ -20,6 +20,19 @@ foo@bar:~$ pip install -r requirements.txt
+1) Установка LibreOffice должна производиться в Docker, иначе при выполнении 
+строки через shell возникнут проблемы с доступом к диреткории /static
+
+2) При запуске автоматических тестирований на Windows необходимо в get_dir.py
+заменить полный путь на вашем компьютере до soffice.exe, если он был установлен нестандартно, который указывается в двойных кавычках 
+**(!!Но затем вернуть всё к исходному!!)** :
```
звучит плохо

mrerro replied:
get_dir.py должен же был как раз убрать этот пункт. По крайней мере заменить его на то что нужно чтобы либре был установлен либо тут, либо тут (https://github.com/profcomff/converter-api/pull/3#discussion_r1161076500) ...

kosedo03 replied:
Добавил проход для шиндоус в Program Files, Program Files (x86). ProgramData и Users

kosedo03 committed:
Добавил инкапсуляцию и почистил

Все работает, красивое...

Temmmmmo commented on file_converter/settings.py:
```diff
@@ -17,6 +18,7 @@ class Config:
         case_sensitive = True
         env_file = ".env"
 
+    ROOT_PATH: str = '/' + os.getenv('APP_NAME', '')
```
`ROOT_PATH: str = '/' + os.getenv('APP_NAME', '')` добавь сюда `| None`

kosedo03 replied:
Ошибку выдает, __or__ не определен для str. Оставлю так.

dyakovri replied:
В типах данных используют именно вертикальную черту

Temmmmmo commented on file_converter/utils/check_pdf.py:
```diff
@@ -1,13 +1,8 @@
+
 async def check_pdf_ok(full_file: str):
     if not exists(full_file):
         return False
```
Сделай короче:
`return exists(full_file)`

Temmmmmo commented on get_dir.py:
```diff
@@ -1,20 +1,47 @@
+    if OS == "Windows":
+        paths = ['\\Program Files', '\\Program Files (x86)', '\\ProgramData', '\\Users']
+        slash = '\\'
+        cd = ext_d[:-2] + slash + 'static' + slash
```
Наверное, лучше f-строку тут сделать, читаемость плохая.

Temmmmmo commented on get_dir.py:
```diff
@@ -1,20 +1,47 @@
+
+    else:
+        slash = '/'
+        cd = ext_d[:-2] + slash + 'static' + slash
```
Аналогично, f-строку

Temmmmmo commented on get_dir.py:
```diff
@@ -1,20 +1,47 @@
+    return get_command(OS, filename)
+
+
+class GetCommand:
+    def __init__(self, filename):
+        f = get_os(filename)
+        self.command = f.get('command')
+        self.direct = f.get('direct')
```
Ты в конструкторе присваиваешь значения полю command и полю direct этого класса. Однако эти поля не определены, добавь их.

kosedo03 replied:
не надо, само инициализирует все хорошо

kosedo03 replied:
исправил, ввел валидацию

Temmmmmo commented on get_dir.py:
```diff
@@ -1,20 +1,47 @@
+    return {'command': command, 'direct': direct}
+
+
+def get_os(filename):
+    OS = platform.system()
+    return get_command(OS, filename)
```
В класс GetCommand лучше добавить в виде метода, вроде ты больше нигде этим не пользуешься.

Temmmmmo commented on get_dir.py:
```diff
@@ -1,20 +1,47 @@
-else:
-    slash = '/'
-    command = 'cd static; libreoffice --headless --convert-to pdf '
+def get_command(OS: str, filename: str):
```
Туда же в класс в в виде метода положи.

kosedo03 replied:
сделал

Temmmmmo commented on file_converter/utils/convertable.py:
```diff
@@ -26,43 +26,44 @@ def get_name(cls) -> str:
 
 def random_str(n):
     alph = string.ascii_letters
     return ''.join(random.choice(alph) for i in range(n))
```
Я думаю, стоит эту функцию в отдельный файл пихнуть.

kosedo03 replied:
пихнул

Temmmmmo reviewed (commented):
Что тут писать? Конец ревью.

kosedo03 committed:
upd

dyakovri commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Выпили папочку и ее в глобальный гитигнор
https://stackoverflow.com/a/7335487/20610418

dyakovri commented on file_converter/settings.py:
```diff
@@ -0,0 +1,31 @@
+        case_sensitive = True
+        env_file = ".env"
+
+    ROOT_PATH: str = '/' + os.getenv('APP_NAME', '')
+    EXTENTIONS: List[str] = ['pdf', 'docx', 'doc']
+    CONVERT_TYPES: List[str] = ['pdf']
+    MAX_SIZE: int = 5000000  # Максимальный размер файла в байтах
+    STATIC_FOLDER: str = DirectoryPath | None
```
Перенеси над конфиг

grigoriev-semyon commented on file_converter/settings.py:
```diff
@@ -0,0 +1,31 @@
+    EXTENTIONS: List[str] = ['pdf', 'docx', 'doc']
+    CONVERT_TYPES: List[str] = ['pdf']
+    MAX_SIZE: int = 5000000  # Максимальный размер файла в байтах
+    STATIC_FOLDER: str = DirectoryPath | None
```
```python 
STATIC_FOLDER: str = DirectoryPath | None
```
это дичь ибо ты присвоил класс в переменную которая должна быть строкового типа
```python 
STATIC_FOLDER:  DirectoryPath | None
```

kosedo03 replied:
заменил на str

grigoriev-semyon commented on file_converter/utils/check_pdf.py:
```diff
@@ -0,0 +1,5 @@
+
+
+async def check_pdf_ok(full_file: str):
+    return exists(full_file)
```
какой смысл в замене одной строки на одну строку? я вижу только минусы

kosedo03 replied:
убрал полностью, теперь напрямую в file

grigoriev-semyon commented on get_dir.py:
```diff
@@ -0,0 +1,54 @@
+            direct = cd + filename
+            command = f'cd {cd}; libreoffice --headless --convert-to pdf {direct}'
+
+        return {'command': command, 'direct': direct}
```
какой смысл запускать его каждый раз?
почему не сделать это при инициализации класса а потом уже отдавать то что есть?

kosedo03 replied:
в подклассах Convertable теперь добавлена функция, также удаление старых файлов перенес в подклассы вместо вызова непосредственно в file_convert

grigoriev-semyon commented on get_dir.py:
```diff
@@ -0,0 +1,54 @@
+import platform
+
+
+class GetCommand:
```
а зачем этот класс? может можно без него? я просил просто переменную прокинуть в Convertable

grigoriev-semyon commented on get_dir.py:
```diff
@@ -0,0 +1,54 @@
+
+    @staticmethod
+    def find(name: str, paths: list):
+        for i in paths:
```
for path in

kosedo03 committed:
upd2

Класс GetCommand теперь создается в convertable.py, импортируем только одну команду

grigoriev-semyon commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+COPY ./${APP_NAME} /app/${APP_NAME}
+
+
+
```
в контейнере нет либры

kosedo03 replied:
Добавил команду

kosedo03 committed:
upd3

kosedo03 committed:
рефактор говнокода

Теперь читаемо!! Сделал собственный класс для понятности 400-й ошибки при конвертации. Проверка успешности конвертации теперь проводится внутри libre.py

grigoriev-semyon commented on file_converter/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+class HTTP_400_BAD_REQUEST(Exception):
```
название дичь

grigoriev-semyon commented on file_converter/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+    def __init__(self):
+        self.message = 'Posted file is corrupted'
+
+    def __str__(self):
+        return f'{self.message}'
```
я думаю такое тут не надо

grigoriev-semyon commented on file_converter/utils/libre.py:
```diff
@@ -0,0 +1,40 @@
+        await run(f'{comm} {filename}')
+        os.remove(f'{cd}{filename}')  # Удаляет старый файл после конвертации
+        if not os.path.exists(f'{cd}{_new_filename}'):  # Проверка на успешность конвертации
+            raise HTTP_400_BAD_REQUEST
```
дичь
я попросил свой класс а не костыль с названием из http))

grigoriev-semyon commented on file_converter/converters/convert.py:
```diff
@@ -0,0 +1,31 @@
+
+    timestamp = str(time.time())
+    _random_str = random_str(10)
+    old_name = f'{timestamp}_{_random_str}.{extension}'
```
почему не _old_name?

grigoriev-semyon commented on file_converter/converters/convert.py:
```diff
@@ -0,0 +1,31 @@
+    timestamp = str(time.time())
+    _random_str = random_str(10)
+    old_name = f'{timestamp}_{_random_str}.{extension}'
+    path = f'{settings.STATIC_FOLDER}/{old_name}'
```
_old_path

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,33 @@
+
+    if to_ext not in settings.CONVERT_TYPES:
+        raise HTTPException(415, 'unsupported to_ext')
+    if file.filename.split(".")[-1] not in settings.EXTENTIONS:
+        raise HTTPException(
+            415,
+            f'Only {", ".join(settings.EXTENTIONS)} files are allowed.',
+        )
```
не будет такой же ошибки как сегодня упала?
если расширение не указано
давай обработку на это и тест

kosedo03 committed:
upd + test

kosedo03 committed:
upd

Переопределил поведение __init__ в исключениях через super()

grigoriev-semyon commented on file_converter/utils/libre.py:
```diff
@@ -0,0 +1,48 @@
+            try:
+                ConverterException()
+            except ConvertError:
+                raise HTTPException(status_code=400, detail='Posted file is corrupted')
```
дичь

grigoriev-semyon commented on file_converter/routes/file.py:
```diff
@@ -0,0 +1,40 @@
+            UnsupportedExt()
+        except ForbiddenExt:
+            raise HTTPException(status_code=415,
+                                detail=f'Only {", ".join(settings.EXTENTIONS)} files are allowed.'
```
дичь

grigoriev-semyon commented on file_converter/utils/libre.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def ConverterException():
+    raise ConvertError
```
дичь

kosedo03 committed:
Больше exceptions не кал

update

kosedo03 committed:
Update file.py

kosedo03 committed:
final_1

Убрал недочеты и дубляж кода

grigoriev-semyon committed:
cicd

grigoriev-semyon committed:
cicd

grigoriev-semyon committed:
isort

grigoriev-semyon committed:
isort

grigoriev-semyon committed:
cicd

grigoriev-semyon committed:
cicd

grigoriev-semyon committed:
fix

grigoriev-semyon committed:
fix

grigoriev-semyon committed:
fix

grigoriev-semyon committed:
fix

grigoriev-semyon committed:
fix

grigoriev-semyon committed:
fix
